#!/bin/bash
del[0]="DROP TABLE IF EXISTS \`analysis\`;"
del[1]="CREATE TABLE \`analysis\` ("
del[2]="  \`analysisId\` int(11) NOT NULL AUTO_INCREMENT,"
del[3]="  \`run_date\` varchar(255) DEFAULT NULL,"
del[4]="  \`logic_name\` varchar(255) NOT NULL,"
del[5]="  \`parameters\` longtext,"
del[6]="  \`assembly_id\` int(11) NOT NULL,"
del[7]="  \`sample_id\` int(11) NOT NULL,"
del[8]="  PRIMARY KEY (\`analysisId\`),"
del[9]="  KEY \`FK_1r72v3obg8nrqab2n52ota3r2\` (\`assembly_id\`),"
del[10]="  KEY \`FK_7qm2opo97g2dosy849bhgsvqb\` (\`sample_id\`),"
del[11]="  CONSTRAINT \`FK_1r72v3obg8nrqab2n52ota3r2\` FOREIGN KEY (\`assembly_id\`) REFERENCES \`assembly\` (\`assembly_id\`),"
del[12]="  CONSTRAINT \`FK_7qm2opo97g2dosy849bhgsvqb\` FOREIGN KEY (\`sample_id\`) REFERENCES \`sample\` (\`sample_id\`)"
del[13]=") ENGINE=InnoDB AUTO_INCREMENT=87 DEFAULT CHARSET=latin1;"
del[14]="DROP TABLE IF EXISTS \`assembly\`;"
del[15]="CREATE TABLE \`assembly\` ("
del[16]="  \`assembly_id\` int(11) NOT NULL AUTO_INCREMENT,"
del[17]="  \`assembly_accession\` varchar(255) NOT NULL,"
del[18]="  \`assembly_name\` varchar(255) NOT NULL,"
del[19]="  \`species_id\` int(11) DEFAULT NULL,"
del[20]="  PRIMARY KEY (\`assembly_id\`),"
del[21]="  UNIQUE KEY \`UK_9ph2yvfhc06pjuwvcjgtxejef\` (\`assembly_accession\`),"
del[22]="  KEY \`FK_okpj44x5yenw18r73xj8px4kf\` (\`species_id\`),"
del[23]="  CONSTRAINT \`FK_okpj44x5yenw18r73xj8px4kf\` FOREIGN KEY (\`species_id\`) REFERENCES \`species\` (\`taxon_id\`)"
del[24]=") ENGINE=InnoDB AUTO_INCREMENT=15 DEFAULT CHARSET=latin1;"
del[25]="DROP TABLE IF EXISTS \`backsplice_junction\`;"
del[26]="CREATE TABLE \`backsplice_junction\` ("
del[27]="  \`junction_id\` int(11) NOT NULL AUTO_INCREMENT,"
del[28]="  \`browser_string\` longtext,"
del[29]="  \`coord_id\` varchar(255) DEFAULT NULL,"
del[30]="  \`genomic_size\` int(11) DEFAULT NULL,"
del[31]="  \`seq_region_end\` int(11) DEFAULT NULL,"
del[32]="  \`seq_region_name\` varchar(255) DEFAULT NULL,"
del[33]="  \`seq_region_start\` int(11) DEFAULT NULL,"
del[34]="  \`seq_region_strand\` varchar(255) DEFAULT NULL,"
del[35]="  \`JPM\` double DEFAULT NULL,"
del[36]="  \`abundance_ratio\` double DEFAULT NULL,"
del[37]="  \`predicted_exons\` varchar(255) DEFAULT NULL,"
del[38]="  \`raw_count\` int(11) DEFAULT NULL,"
del[39]="  \`splice_signal\` varchar(255) DEFAULT NULL,"
del[40]="  \`analysis_id\` int(11) NOT NULL,"
del[41]="  \`locus_id\` int(11) DEFAULT NULL,"
del[42]="  \`classification\` varchar(255) DEFAULT NULL,"
del[43]="  \`in_platelets\` bit(1) DEFAULT NULL,"
del[44]="  \`is_published\` bit(1) DEFAULT NULL,"
del[45]="  \`junction_repeat_coverage\` double DEFAULT NULL,"
del[46]="  \`n_methods\` int(11) DEFAULT NULL,"
del[47]="  \`splice_type\` varchar(255) DEFAULT NULL,"
del[48]="  \`spliced_size\` int(11) DEFAULT NULL,"
del[49]="  \`splice_3_support\` int(11) DEFAULT NULL,"
del[50]="  \`splice_5_support\` int(11) DEFAULT NULL,"
del[51]="  \`splice_site_count\` int(11) DEFAULT NULL,"
del[52]="  \`TPM\` double DEFAULT NULL,"
del[53]="  \`transcript_count\` double DEFAULT NULL,"
del[54]="  \`gc_perc\` double DEFAULT NULL,"
del[55]="  PRIMARY KEY (\`junction_id\`),"
del[56]="  KEY \`FK_7g3afyllcsldk5t9nd012hu93\` (\`locus_id\`),"
del[57]="  KEY \`FK_gdppp9c4041kpnv1wocvscs5j\` (\`analysis_id\`),"
del[58]="  CONSTRAINT \`FK_7g3afyllcsldk5t9nd012hu93\` FOREIGN KEY (\`locus_id\`) REFERENCES \`locus\` (\`locusId\`),"
del[59]="  CONSTRAINT \`FK_gdppp9c4041kpnv1wocvscs5j\` FOREIGN KEY (\`analysis_id\`) REFERENCES \`analysis\` (\`analysisId\`)"
del[60]=") ENGINE=InnoDB AUTO_INCREMENT=300895 DEFAULT CHARSET=latin1;"
del[61]="DROP TABLE IF EXISTS \`canonical_junction\`;"
del[62]="CREATE TABLE \`canonical_junction\` ("
del[63]="  \`junction_id\` int(11) NOT NULL AUTO_INCREMENT,"
del[64]="  \`browser_string\` longtext,"
del[65]="  \`coord_id\` varchar(255) DEFAULT NULL,"
del[66]="  \`genomic_size\` int(11) DEFAULT NULL,"
del[67]="  \`seq_region_end\` int(11) DEFAULT NULL,"
del[68]="  \`seq_region_name\` varchar(255) DEFAULT NULL,"
del[69]="  \`seq_region_start\` int(11) DEFAULT NULL,"
del[70]="  \`seq_region_strand\` varchar(255) DEFAULT NULL,"
del[71]="  \`JPM\` double DEFAULT NULL,"
del[72]="  \`predicted_exons\` varchar(255) DEFAULT NULL,"
del[73]="  \`raw_count\` int(11) DEFAULT NULL,"
del[74]="  \`splice_signal\` varchar(255) DEFAULT NULL,"
del[75]="  \`analysis_id\` int(11) NOT NULL,"
del[76]="  \`locus_id\` int(11) DEFAULT NULL,"
del[77]="  \`splice_type\` varchar(255) DEFAULT NULL,"
del[78]="  PRIMARY KEY (\`junction_id\`),"
del[79]="  KEY \`FK_8lpr7o5u75mklseeyamja334k\` (\`locus_id\`),"
del[80]="  KEY \`FK_lui5gi1tudm2da24k61bhfw3t\` (\`analysis_id\`),"
del[81]="  CONSTRAINT \`FK_8lpr7o5u75mklseeyamja334k\` FOREIGN KEY (\`locus_id\`) REFERENCES \`locus\` (\`locusId\`),"
del[82]="  CONSTRAINT \`FK_lui5gi1tudm2da24k61bhfw3t\` FOREIGN KEY (\`analysis_id\`) REFERENCES \`analysis\` (\`analysisId\`)"
del[83]=") ENGINE=InnoDB AUTO_INCREMENT=27199799 DEFAULT CHARSET=latin1;"
del[84]="DROP TABLE IF EXISTS \`exon\`;"
del[85]="CREATE TABLE \`exon\` ("
del[86]="  \`locus_id\` int(11) NOT NULL,"
del[87]="  \`coord_id\` varchar(255) DEFAULT NULL,"
del[88]="  \`genomic_size\` int(11) DEFAULT NULL,"
del[89]="  \`seq_region_end\` int(11) DEFAULT NULL,"
del[90]="  \`seq_region_name\` varchar(255) DEFAULT NULL,"
del[91]="  \`seq_region_start\` int(11) DEFAULT NULL,"
del[92]="  \`seq_region_strand\` varchar(255) DEFAULT NULL,"
del[93]="  \`rank\` int(11) DEFAULT NULL,"
del[94]="  \`stable_id\` varchar(255) DEFAULT NULL,"
del[95]="  KEY \`FK_rn2bqjh4aik56jnrpagwh2br0\` (\`locus_id\`),"
del[96]="  CONSTRAINT \`FK_rn2bqjh4aik56jnrpagwh2br0\` FOREIGN KEY (\`locus_id\`) REFERENCES \`locus\` (\`locusId\`)"
del[97]=") ENGINE=InnoDB DEFAULT CHARSET=latin1;"
del[98]="DROP TABLE IF EXISTS \`hibernate_sequences\`;"
del[99]="CREATE TABLE \`hibernate_sequences\` ("
del[100]="  \`sequence_name\` varchar(255) DEFAULT NULL,"
del[101]="  \`sequence_next_hi_value\` int(11) DEFAULT NULL"
del[102]=") ENGINE=InnoDB DEFAULT CHARSET=latin1;"
del[103]="DROP TABLE IF EXISTS \`locus\`;"
del[104]="CREATE TABLE \`locus\` ("
del[105]="  \`locusId\` int(11) NOT NULL AUTO_INCREMENT,"
del[106]="  \`browser_string\` longtext,"
del[107]="  \`circrna_abundance_ratio\` double DEFAULT NULL,"
del[108]="  \`coord_id\` varchar(255) DEFAULT NULL,"
del[109]="  \`genomic_size\` int(11) DEFAULT NULL,"
del[110]="  \`seq_region_end\` int(11) DEFAULT NULL,"
del[111]="  \`seq_region_name\` varchar(255) DEFAULT NULL,"
del[112]="  \`seq_region_start\` int(11) DEFAULT NULL,"
del[113]="  \`seq_region_strand\` varchar(255) DEFAULT NULL,"
del[114]="  \`gene_name\` varchar(255) DEFAULT NULL,"
del[115]="  \`is_circrna_host\` bit(1) DEFAULT NULL,"
del[116]="  \`nexons\` int(11) DEFAULT NULL,"
del[117]="  \`source\` varchar(255) DEFAULT NULL,"
del[118]="  \`spliced_size\` int(11) DEFAULT NULL,"
del[119]="  \`stable_id\` varchar(255) DEFAULT NULL,"
del[120]="  \`total_backsplice_reads\` double DEFAULT NULL,"
del[121]="  \`total_splice_reads\` double DEFAULT NULL,"
del[122]="  \`assembly_id\` int(11) NOT NULL,"
del[123]="  PRIMARY KEY (\`locusId\`),"
del[124]="  KEY \`FK_kaidlko80rodyo94ulvw9dklf\` (\`assembly_id\`),"
del[125]="  CONSTRAINT \`FK_kaidlko80rodyo94ulvw9dklf\` FOREIGN KEY (\`assembly_id\`) REFERENCES \`assembly\` (\`assembly_id\`)"
del[126]=") ENGINE=InnoDB AUTO_INCREMENT=261163 DEFAULT CHARSET=latin1;"
del[127]="DROP TABLE IF EXISTS \`locus_expression\`;"
del[128]="CREATE TABLE \`locus_expression\` ("
del[129]="  \`expression_id\` int(11) NOT NULL AUTO_INCREMENT,"
del[130]="  \`deseq_normalised\` double DEFAULT NULL,"
del[131]="  \`FPKM\` double DEFAULT NULL,"
del[132]="  \`RPKM\` double DEFAULT NULL,"
del[133]="  \`rpkm_external\` double DEFAULT NULL,"
del[134]="  \`rpkm_internal\` double DEFAULT NULL,"
del[135]="  \`rpkm_ratio\` double DEFAULT NULL,"
del[136]="  \`TPM\` double DEFAULT NULL,"
del[137]="  \`analysis_id\` int(11) DEFAULT NULL,"
del[138]="  \`locus_id\` int(11) DEFAULT NULL,"
del[139]="  PRIMARY KEY (\`expression_id\`),"
del[140]="  KEY \`FK_1d965wkkt5my90lgewvafu43x\` (\`analysis_id\`),"
del[141]="  KEY \`FK_4vef6i1xkpngahbsf3jknsypt\` (\`locus_id\`),"
del[142]="  CONSTRAINT \`FK_1d965wkkt5my90lgewvafu43x\` FOREIGN KEY (\`analysis_id\`) REFERENCES \`analysis\` (\`analysisId\`),"
del[143]="  CONSTRAINT \`FK_4vef6i1xkpngahbsf3jknsypt\` FOREIGN KEY (\`locus_id\`) REFERENCES \`locus\` (\`locusId\`)"
del[144]=") ENGINE=InnoDB DEFAULT CHARSET=latin1;"
del[145]="DROP TABLE IF EXISTS \`sample\`;"
del[146]="CREATE TABLE \`sample\` ("
del[147]="  \`sample_id\` int(11) NOT NULL AUTO_INCREMENT,"
del[148]="  \`accession\` varchar(50) NOT NULL,"
del[149]="  \`backspliced_reads\` double DEFAULT NULL,"
del[150]="  \`chimeric_reads\` double DEFAULT NULL,"
del[151]="  \`circrna_count\` int(11) DEFAULT NULL,"
del[152]="  \`description\` longtext,"
del[153]="  \`fastqc_path\` varchar(255) DEFAULT NULL,"
del[154]="  \`ftp_path\` varchar(255) DEFAULT NULL,"
del[155]="  \`instrument\` varchar(100) DEFAULT NULL,"
del[156]="  \`library_size\` int(11) DEFAULT NULL,"
del[157]="  \`library_strategy\` varchar(255) DEFAULT NULL,"
del[158]="  \`gtag_cjuncs_reads\` double DEFAULT NULL,"
del[159]="  \`mapped_reads\` double DEFAULT NULL,"
del[160]="  \`read_length\` int(11) DEFAULT NULL,"
del[161]="  \`fraction\` varchar(100) DEFAULT NULL,"
del[162]="  \`submitter\` varchar(100) DEFAULT NULL,"
del[163]="  \`source\` varchar(50) DEFAULT NULL,"
del[164]="  \`total_spliced_reads\` double DEFAULT NULL,"
del[165]="  \`source_id\` int(11) DEFAULT NULL,"
del[166]="  PRIMARY KEY (\`sample_id\`),"
del[167]="  UNIQUE KEY \`UK_7ntygu84trwg1q88x2doa8m8e\` (\`accession\`),"
del[168]="  KEY \`FK_ffihengafhkia0btjoyrpptih\` (\`species_id\`),"
del[169]="  CONSTRAINT \`FK_ffihengafhkia0btjoyrpptih\` FOREIGN KEY (\`species_id\`) REFERENCES \`species\` (\`taxon_id\`)"
del[170]=") ENGINE=InnoDB AUTO_INCREMENT=10 DEFAULT CHARSET=latin1;"
del[171]="DROP TABLE IF EXISTS \`species\`;"
del[172]="CREATE TABLE \`species\` ("
del[173]="  \`taxon_id\` int(11) NOT NULL,"
del[174]="  \`species_name\` varchar(255) DEFAULT NULL,"
del[175]="  PRIMARY KEY (\`taxon_id\`)"
del[176]=") ENGINE=InnoDB DEFAULT CHARSET=latin1;"
del[177]="DROP TABLE IF EXISTS \`ensembl_gene\`;"
del[178]="CREATE TABLE \`ensembl_gene\` ("
del[179]="  \`gene_id\` int(11) NOT NULL AUTO_INCREMENT,"
del[180]="  \`coord_id\` varchar(255) DEFAULT NULL,"
del[181]="  \`seq_region_end\` int(11) DEFAULT NULL,"
del[182]="  \`seq_region_name\` varchar(255) DEFAULT NULL,"
del[183]="  \`seq_region_start\` int(11) DEFAULT NULL,"
del[184]="  \`seq_region_strand\` varchar(255) DEFAULT NULL,"
del[185]="  \`gene_name\` varchar(255) DEFAULT NULL,"
del[186]="  \`description\` varchar(255) DEFAULT NULL,"
del[187]="  \`is_circrna_host\` bit(1) DEFAULT NULL,"
del[188]="  \`biotype\` varchar(255) DEFAULT NULL,"
del[189]="  \`stable_id\` varchar(255) DEFAULT NULL,"
del[190]="  \`assembly_id\` int(11) NOT NULL,"
del[191]="  PRIMARY KEY (\`gene_id\`),"
del[192]="  KEY \`assembly_id\` (\`assembly_id\`),"
del[193]="  CONSTRAINT \`ensembl_gene_ibfk_1\` FOREIGN KEY (\`assembly_id\`) REFERENCES \`assembly\` (\`assembly_id\`)"
del[194]=") ENGINE=InnoDB AUTO_INCREMENT=261164 DEFAULT CHARSET=latin1;"
del[195]="LOCK TABLES \`ensembl_gene\` WRITE;"
del[196]=") ENGINE=InnoDB AUTO_INCREMENT=729022 DEFAULT CHARSET=latin1;"
del[197]="  \`bigwig_path\` varchar(255) DEFAULT NULL,"
del[198]="  \`project\` varchar(255) DEFAULT NULL,"
del[199]="  \`TIN\` double DEFAULT NULL,"
del[200]=") ENGINE=InnoDB AUTO_INCREMENT=96 DEFAULT CHARSET=latin1;"
del[201]=") ENGINE=InnoDB AUTO_INCREMENT=5180574 DEFAULT CHARSET=latin1;"
del[202]="  \`raw_count\` double NOT NULL,"

addcore[0]="\`analysis\`"
addcore[1]="\`assembly\`"
addcore[2]="\`backsplice_junction\`"
addcore[3]="\`canonical_junction\`"
addcore[4]="\`exon\`"
addcore[5]="\`locus\`"
addcore[6]="\`locus_expression\`"
addcore[7]="\`sample\`"
addcore[8]="\`species\`"
addcore[9]="\`ensembl_gene\`"
addcore[10]="\`hibernate_sequences\`"

addedcore[0]="\`core_analysis\`"
addedcore[1]="\`core_assembly\`"
addedcore[2]="\`core_backsplicejunction\`"
addedcore[3]="\`core_canonicaljunction\`"
addedcore[4]="\`core_exon\`"
addedcore[5]="\`core_locus\`"
addedcore[6]="\`core_locusexpression\`"
addedcore[7]="\`core_sample\`"
addedcore[8]="\`core_species\`"
addedcore[9]="\`core_ensemblgene\`"
addedcore[10]="\`core_hibernatesequences\`"

tablemissingfields[0]="INSERT INTO \`core_analysis\` VALUES"
tablemissingfields[1]="INSERT INTO \`core_assembly\` VALUES"
tablemissingfields[2]="INSERT INTO \`core_backsplicejunction\` VALUES"
tablemissingfields[3]="INSERT INTO \`core_canonicaljunction\` VALUES"
tablemissingfields[4]="INSERT INTO \`core_exon\` VALUES"
tablemissingfields[5]="INSERT INTO \`core_locus\` VALUES"
tablemissingfields[6]="INSERT INTO \`core_locusexpression\` VALUES"
tablemissingfields[7]="INSERT INTO \`core_sample\` VALUES"
tablemissingfields[8]="INSERT INTO \`core_species\` VALUES"
tablemissingfields[9]="INSERT INTO \`core_ensemblgene\` VALUES"
tablemissingfields[10]="INSERT INTO \`core_hibernatesequences\` VALUES"

tableaddedfields[0]="INSERT INTO \`core_analysis\` (analysis_id, run_date, logic_name, parameters, assembly_id_id, sample_id_id) VALUES"
tableaddedfields[1]="INSERT INTO \`core_assembly\` (assembly_id, assembly_accession, assembly_name, species_id_id) VALUES"
tableaddedfields[2]="INSERT INTO \`core_backsplicejunction\` (junction_id, browser_string, coord_id, genomic_size, seq_region_end, seq_region_name, seq_region_start, seq_region_strand, jpm, abundance_ratio, predicted_exons, raw_count, splice_signal, analysis_id_id, locus_id_id, classification, in_platelets, is_published, junction_repeat_coverage, n_methods, splice_type, spliced_size, splice_3_support, splice_5_support, splice_site_count, tpm, transcript_count, gc_perc) VALUES"
tableaddedfields[3]="INSERT INTO \`core_canonicaljunction\` (junction_id, browser_string, coord_id, genomic_size, seq_region_end, seq_region_name, seq_region_start, seq_region_strand, jpm, predicted_exons, raw_count, splice_signal, analysis_id_id, locus_id_id, splice_type) VALUES"
tableaddedfields[4]="INSERT INTO \`core_exon\` (locus_id, coord_id, genomic_size, seq_region_end, seq_region_name, seq_region_start, seq_region_strand, rank, stable_id) VALUES"
tableaddedfields[5]="INSERT INTO \`core_locus\` (locus_id, browser_string, coord_id, genomic_size, seq_region_end, seq_region_name, seq_region_start, seq_region_strand, gene_name, is_circrna_host, nexons, source, spliced_size, stable_id, total_backsplice_reads, total_splice_reads, assembly_id_id) VALUES"
tableaddedfields[6]="INSERT INTO \`core_locusexpression\` (expression_id, raw_count, rpkm, rpkm_external, rpkm_internal, rpkm_ratio, tpm, analysis_id_id, locus_id_id) VALUES"
tableaddedfields[7]="INSERT INTO \`core_sample\` (sample_id, accession, backspliced_reads, chimeric_reads, circrna_count, description, fastq_path, bigwig_path, ftp_path, instrument, library_size, library_strategy, project, gtag_cjuncs_reads, mapped_reads, read_length, fraction, submitter, source, total_spliced_reads, TIN, species_id_id) VALUES"
tableaddedfields[8]="INSERT INTO \`core_species\` (taxon_id, scientific_name) VALUES"
tableaddedfields[9]="INSERT INTO \`core_ensemblgene\` (gene_id, coord_id, seq_region_end, seq_region_name, seq_region_start, seq_region_strand, gene_name, description, is_circrna_host, biotype, stable_id, assembly_id_id) VALUES"
tableaddedfields[10]="INSERT INTO \`core_hibernatesequences\` (sequence_name, sequence_next_hi_value) VALUES"


# Delete create table statements
for index in ${!del[*]}
do
    sed -i "s|${del[$index]}.*||g" db.txt
    echo $index
done

# Change table names
for index in ${!addcore[*]}
do
    sed -i "s|${addcore[$index]}|${addedcore[$index]}|" db.txt
    echo $((index + 201))
done

# Change table field names
for index in ${!tablemissingfields[*]}
do
    sed -i "s|${tablemissingfields[$index]}|${tableaddedfields[$index]}|" db.txt
    echo $((index + 210))
done
